client/network/src/discovery.rs

```diff
@@ -483,29 +483,86 @@ pub enum DiscoveryOut {
 }
 
 impl NetworkBehaviour for DiscoveryBehaviour {
-	type ConnectionHandler = ToggleIntoConnectionHandler<KademliaHandlerProto<QueryId>>;
+	type ConnectionHandler = ToggleConnectionHandler<KademliaHandler<QueryId>>;
 	type OutEvent = DiscoveryOut;
 
-	fn new_handler(&mut self) -> Self::ConnectionHandler {
-		self.kademlia.new_handler()
+	fn handle_established_inbound_connection(
+		&mut self,
+		connection_id: ConnectionId,
+		peer: PeerId,
+		local_addr: &Multiaddr,
+		remote_addr: &Multiaddr,
+	) -> Result<THandler<Self>, ConnectionDenied> {
+		self.kademlia.handle_established_inbound_connection(
+			connection_id,
+			peer,
+			local_addr,
+			remote_addr,
+		)
 	}
 
-	fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
-		let mut list = self
-			.permanent_addresses
-			.iter()
-			.filter_map(|(p, a)| if p == peer_id { Some(a.clone()) } else { None })
-			.collect::<Vec<_>>();
+	fn handle_established_outbound_connection(
+		&mut self,
+		connection_id: ConnectionId,
+		peer: PeerId,
+		addr: &Multiaddr,
+		role_override: Endpoint,
+	) -> Result<THandler<Self>, ConnectionDenied> {
+		self.kademlia.handle_established_outbound_connection(
+			connection_id,
+			peer,
+			addr,
+			role_override,
+		)
+	}
+
+	fn handle_pending_inbound_connection(
+		&mut self,
+		connection_id: ConnectionId,
+		local_addr: &Multiaddr,
+		remote_addr: &Multiaddr,
+	) -> Result<(), ConnectionDenied> {
+		self.kademlia
+			.handle_pending_inbound_connection(connection_id, local_addr, remote_addr)
+	}
+
+	fn handle_pending_outbound_connection(
+		&mut self,
+		connection_id: ConnectionId,
+		maybe_peer: Option<PeerId>,
+		addresses: &[Multiaddr],
+		effective_role: Endpoint,
+	) -> Result<Vec<Multiaddr>, ConnectionDenied> {
+		let mut list = Vec::new();
+
+		if let Some(peer_id) = maybe_peer {
+			list.extend(
+				self.permanent_addresses
+					.iter()
+					.filter_map(|(p, a)| if *p == peer_id { Some(a.clone()) } else { None })
+					.collect::<Vec<_>>(),
+			);
 
-		if let Some(ephemeral_addresses) = self.ephemeral_addresses.get(peer_id) {
-			list.extend(ephemeral_addresses.clone());
+			if let Some(ephemeral_addresses) = self.ephemeral_addresses.get(&peer_id) {
+				list.extend(ephemeral_addresses.clone());
+			}
 		}
 
 		{
-			let mut list_to_filter = self.kademlia.addresses_of_peer(peer_id);
+			let mut list_to_filter = self.kademlia.handle_pending_outbound_connection(
```

Comment:
should we attempt to get more addresses from kademlia and mdns for a peer if there's no ID (`maybe_peer` is `None`)?

Reply:
Is there easy way to do it? I'd probably not bother with it and just let the next random walk find new peers.

Reply:
> Is there easy way to do it?

the code already does this. no need to do anything.

I was asking if it even makes sense to probe kademlia / mdns if peerID is empty. I guess it won't hurt. Otherwise we can just return empty `Vec` if `maybe_peer.is_none()` in the beginning of this function.

Reply:
I would err on the side of caution and keep the old behavior for now at least.

Reply:
Addressed in https://github.com/paritytech/substrate/pull/13587/commits/a02f891a2af54fba1dfc0022aed9a7fd85b98a3e